src/v/utils/xid.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/seastarx.h"
+#include "bytes/iobuf_parser.h"
+
+#include <seastar/core/sstring.hh>
+
+#include <fmt/core.h>
+
+#include <array>
+#include <cstdint>
+#include <exception>
+
+/**
+ * Exception indicating that provided XID isn't valid f.e. incorrect length of
+ * invalid characters
+ */
+class invalid_xid final : public std::exception {
+public:
+    explicit invalid_xid(const ss::sstring&);
+    const char* what() const noexcept final { return _msg.c_str(); }
+
+private:
+    ss::sstring _msg;
+};
+
+/**
+ * Class representing an XID.
+ *
+ * The binary representation of the id has 12 bytes. String representation is
+ * using base32 encoding (w/o padding) for better space efficiency when stored
+ * in that form (20 bytes).
+ *
+ * Based on (https://github.com/rs/xid)
+ */
+class xid {
```

Comment:
i wonder if all namespace-topic-partitions should just be xids now ? maybe the flywheight pattern travis had in mind is better for partition ids (ints are cheaper)

Reply:
this is good idea, it would require a lot of changes, we have numeric ntps planned.